### DISCUSSION TITLE
How to use static IP stored in dynamic config.

### DISCUSSION THREAD

**Discussion by bkrajendra** [khoih-prog/WiFiManager_RTL8720#4] [Q&A]
I need an example of using static IP from dynamic parameters.
## Questions:
How to use dynamic parameters in code.?
At what point are the values available to use?
in the case of a static IP address, it will be required at the wifi connection time. How to use dynamic parameter from config form to assign it to ```WiFiManager->setSTAStaticIPConfig(static_ip);``` ?

@khoih-prog  Please help

**Comment by khoih-prog** [comment]
It's good that you've spent some time to read the example code and know we can use `WiFiManager->setSTAStaticIPConfig(static_ip);`

---

> At what point are the values available to use?

After the first call to `getConfigData()` in library code, which happens in `begin()`

https://github.com/khoih-prog/WiFiManager_RTL8720/blob/9659ee5af5f2709c5d59d9db184b2b2dcbd4b077/examples/RTL8720_WiFi/RTL8720_WiFi.ino#L113

---

To get the `Config Data` anytime, before of after `begin()`, use

https://github.com/khoih-prog/WiFiManager_RTL8720/blob/9659ee5af5f2709c5d59d9db184b2b2dcbd4b077/src/WiFiManager_RTL8720.h#L699

then `Config Data` is ready to be used

You really don't need to do this after the following `begin()`

https://github.com/khoih-prog/WiFiManager_RTL8720/blob/9659ee5af5f2709c5d59d9db184b2b2dcbd4b077/examples/RTL8720_WiFi/RTL8720_WiFi.ino#L113

---

> How to use dynamic parameters in code.?

See example

https://github.com/khoih-prog/WiFiManager_RTL8720/blob/9659ee5af5f2709c5d59d9db184b2b2dcbd4b077/examples/RTL8720_WiFi/RTL8720_WiFi.ino#L118-L153

---

> I need an example of using static IP from dynamic parameters.

> How to use dynamic parameter from config form to assign it to WiFiManager->setSTAStaticIPConfig(static_ip); ?

This is left as an exercise for you, after the above answers.

If not possible, it'll be a serious problem, and you have to post on the Arduino Forum for help. I hope I have time to deal with all the basic knowledge and coding and help every single person.

**Comment by bkrajendra** [comment]
@khoih-prog thanks for the answer.
This is what I have done, and its working fine. let me know if I am doing overkill for string to IP, or is there any native method to do that.?

```c
void Setup(){
WiFiManager->begin(HOST_NAME);
  String mip = myMenuItems[0].pdata;

  // char item[4] = strtok(mip, "."); // this did not worked

  IPAddress static_IP = str2IP(mip);
  WiFiManager->setSTAStaticIPConfig(static_IP);
  Serial.println("Static IP menu \r\n");
  Serial.println(myMenuItems[0].pdata);

}

IPAddress str2IP(String str) {
  IPAddress ret(getIpBlock(0, str), getIpBlock(1, str), getIpBlock(2, str), getIpBlock(3, str));
  return ret;
}

int getIpBlock(int index, String str) {
  char separator = '.';
  int found = 0;
  int strIndex[] = { 0, -1 };
  int maxIndex = str.length() - 1;

  for (int i = 0; i <= maxIndex && found <= index; i++) {
    if (str.charAt(i) == separator || i == maxIndex) {
      found++;
      strIndex[0] = strIndex[1] + 1;
      strIndex[1] = (i == maxIndex) ? i + 1 : i;
    }
  }
  return found > index ? str.substring(strIndex[0], strIndex[1]).toInt() : 0;
}
```

I have one more question. Is there any way to save/change the config data (one or multiple values) after the wifi connection? I want to have rest API to do that.

**Reply by khoih-prog** [reply]
Now, the question has no more relation to this library. I suggest you have to study / research yourself, or post it on other Forums, such as  Arduino Forum.

**Reply by bkrajendra** [reply]
anyways thank you very much for the support.
Just about the last question, it's my mistake I added multiple things together. The question is only related to the library.

Q: How to update DynamiData after the wifi connection success?

I have seen the function FlashStorage_putDynamicData(). I'll play around with this and figure it out.

Thanks again.
